File: scripts/benchmark.mjs
import { performance } from 'node:perf_hooks';
import { mkdirSync, readFileSync, readdirSync, writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { AssuranceEngine } from '@l9/assurance-cli';
import { admitObservations, validateObservation } from '@l9/assurance-evidence';
import { renderDecisionSummary, verifyDecision } from '@l9/assurance-evaluator';
import { ROOT } from './lib/files.mjs';

const FIXED_TIME = '2026-07-21T00:00:02.000Z';
const outputIndex = process.argv.indexOf('--output');
const outputPath = outputIndex >= 0 ? resolve(process.argv[outputIndex + 1] ?? '') : undefined;

function readJson(...parts) {
  return JSON.parse(readFileSync(join(ROOT, ...parts), 'utf8'));
}

function configuration() {
  const controls = readdirSync(join(ROOT, 'controls', 'ci'))
    .filter((name) => name.endsWith('.yaml'))
    .sort()
    .map((name) => readJson('controls', 'ci', name));
  return {
    producerRegistry: readJson('fixtures', 'compatibility', 'producer-registry.trusted.json'),
    checkRegistry: readJson('fixtures', 'compatibility', 'check-registry.json'),
    profile: readJson('fixtures', 'compatibility', 'profile.json'),
    policy: readJson('fixtures', 'compatibility', 'policy.json'),
    controls,
    clock: () => FIXED_TIME,
    evaluator: {
      id: 'l9-assurance',
      version: '2.0.1',
      repository: 'Quantum-L9/l9-assurance',
    },
  };
}

function percentile(values, quantile) {
  const ordered = [...values].sort((left, right) => left - right);
  return ordered[Math.min(ordered.length - 1, Math.floor(ordered.length * quantile))] ?? 0;
}

const subject = readJson('fixtures', 'valid', 'subject.json');
const baseObservation = readJson('fixtures', 'valid', 'tests.observation.json');
const validationDurations = [];
for (let index = 0; index < 200; index += 1) {
  const started = performance.now();
  const validation = validateObservation(baseObservation);
  validationDurations.push(performance.now() - started);
  if (!validation.valid) throw new Error('Benchmark fixture failed structural validation.');
}
const validationP95Ms = percentile(validationDurations, 0.95);

const observations = Array.from({ length: 1000 }, (_, index) => ({
  ...baseObservation,
  observationId: `obs_benchmark_${index}`,
  execution: { ...baseObservation.execution, runId: `run_benchmark_${index}` },
}));
const config = configuration();
const admissionStarted = performance.now();
const admission = await admitObservations(observations, {
  subject,
  producerRegistry: config.producerRegistry,
  checkRegistry: config.checkRegistry,
  receivedAt: FIXED_TIME,
  channel: 'local',
});
const admissionMs = performance.now() - admissionStarted;
if (admission.accepted.length !== 1000) throw new Error('Benchmark did not admit all observations.');
const rssBytesAfterAdmission = process.memoryUsage().rss;

const singleEngine = new AssuranceEngine(config);
const singleAdmission = await singleEngine.admit({
  subject,
  observations: [baseObservation],
  receivedAt: FIXED_TIME,
});
const baseControl = readJson('controls', 'ci', 'tests.yaml');
const controls = Array.from({ length: 500 }, (_, index) => {
  const suffix = String(index).padStart(3, '0');
  return {
    ...baseControl,
    id: `L9.BENCHMARK.CONTROL_${suffix}`,
    claim: `l9.claim.benchmark-${suffix}`,
    title: `Benchmark control ${suffix}`,
    description: `Synthetic benchmark control ${suffix}.`,
  };
});
const benchmarkEngine = new AssuranceEngine({
  ...config,
  controls,
  profile: {
    ...config.profile,
    id: 'l9.benchmark',
    controls: controls.map((control) => ({ id: control.id, version: control.version })),
    outputClaims: [{ id: 'l9.claim.benchmark-satisfied', version: '1.0.0' }],
  },
});
const evaluationStarted = performance.now();
const decision = await benchmarkEngine.evaluate({
  subject,
  acceptedEvidence: singleAdmission.accepted,
  evaluationTime: FIXED_TIME,
  decisionId: 'dec_benchmark_500_controls',
});
const evaluationMs = performance.now() - evaluationStarted;

const verificationStarted = performance.now();
const verification = verifyDecision(decision);
const verificationMs = performance.now() - verificationStarted;
if (!verification.valid) throw new Error(`Benchmark decision verification failed: ${verification.reasons.join('; ')}`);

const summaryStarted = performance.now();
renderDecisionSummary(decision);
const summaryMs = performance.now() - summaryStarted;

const targets = {
  validationP95Ms: 100,
  admission1000Ms: 5000,
  evaluation500ControlsMs: 2000,
  verificationMs: 250,
  summaryMs: 500,
  rssBytesAfterAdmission: 512 * 1024 * 1024,
};
const measurements = {
  validationP95Ms,
  admission1000Ms: admissionMs,
  evaluation500ControlsMs: evaluationMs,
  verificationMs,
  summaryMs,
  rssBytesAfterAdmission,
};
const checks = Object.fromEntries(
  Object.entries(targets).map(([name, target]) => [name, measurements[name] < target]),
);
const report = {
  schema: 'l9.assurance-benchmark',
  schemaVersion: '1.0.0',
  executedAt: new Date().toISOString(),
  runtime: { node: process.version, platform: process.platform, architecture: process.arch },
  measurements,
  targets,
  checks,
  passed: Object.values(checks).every(Boolean),
};

const serialized = `${JSON.stringify(report, null, 2)}\n`;
if (outputPath) {
  mkdirSync(dirname(outputPath), { recursive: true });
  writeFileSync(outputPath, serialized, 'utf8');
}
process.stdout.write(serialized);
if (!report.passed) process.exitCode = 1;

File: scripts/benchmark.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from time import perf_counter

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "src"))

from l9_assurance.cli import AssuranceEngine, load_configuration  # noqa: E402
from l9_assurance.testing import build_observation, build_subject  # noqa: E402


def build_report() -> dict[str, object]:
    config = load_configuration()
    config["producerRegistry"] = json.loads(
        (ROOT / "fixtures" / "compatibility" / "producer-registry.trusted.json").read_text(
            encoding="utf-8"
        )
    )
    config["checkRegistry"] = json.loads(
        (ROOT / "fixtures" / "compatibility" / "check-registry.json").read_text(encoding="utf-8")
    )
    engine = AssuranceEngine(config)
    subject = build_subject()
    observations = [
        build_observation(observation_id=f"obs_{index}", subject=subject) for index in range(1_000)
    ]

    started = perf_counter()
    admission = engine.admit(
        subject,
        observations,
        received_at="2026-07-21T00:00:02Z",
    )
    admission_seconds = perf_counter() - started

    started = perf_counter()
    for index in range(500):
        engine.evaluate(
            subject,
            admission["accepted"],
            evaluation_time="2026-07-21T00:00:03Z",
            decision_id=f"dec_bench_{index}",
        )
    evaluation_seconds = perf_counter() - started

    return {
        "schema": "l9.assurance-benchmark",
        "schemaVersion": "1.0.0",
        "admission": {
            "observations": 1_000,
            "seconds": round(admission_seconds, 6),
            "objectiveSeconds": 5.0,
            "passed": admission_seconds < 5.0,
        },
        "evaluation": {
            "decisions": 500,
            "seconds": round(evaluation_seconds, 6),
            "objectiveSeconds": 2.0,
            "passed": evaluation_seconds < 2.0,
        },
    }


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run deterministic Assurance performance objectives."
    )
    parser.add_argument("--output", type=Path, help="Optional report destination.")
    args = parser.parse_args()

    report = build_report()
    text = json.dumps(report, indent=2) + "\n"
    if args.output is not None:
        destination = args.output if args.output.is_absolute() else ROOT / args.output
        destination.parent.mkdir(parents=True, exist_ok=True)
        destination.write_text(text, encoding="utf-8")
    print(text, end="")
    return 0 if report["admission"]["passed"] and report["evaluation"]["passed"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
